### DISCUSSION TITLE
VSCode extension

### DISCUSSION THREAD

**Discussion by pokey** [arxanas/git-branchless#567] [Ideas]
It would be useful to have a VSCode extension for `git-branchless`.  Here are a couple feature ideas:

- Webview with live-updating smart log
- VSCode commands corresponding to various cli commands
- Might be cool to have a GUI-based commit graph manipulator including eg drag and drop
- Syntax highlighting for query language, esp as it grows (eg https://github.com/arxanas/git-branchless/discussions/568)

**Comment by pokey** [comment]
I'm considering throwing together a super simple POC, primarily for personal use, that is just a really thin wrapper around my most common `git-branchless` operations

**Comment by arxanas** [comment]
Facebook used to have a tool like this (for Mercurial) called the Interactive Smartlog, which was [really good](https://twitter.com/threepointone). Unfortunately, there's not even any documentation/pictures about it online. It supported drag/drop for commits, and, among other things, it also had a really good interactive changes selector for commits. The closest thing I've found externally is [GitUp](https://gitup.co/), which you might want to check out.

**Reply by arxanas** [reply]
Update: Facebook has since published the Interactive Smartlog via Sapling: https://sapling-scm.com/docs/addons/isl/. It's not up to parity with their previous internal version yet, but hopefully it will get there soon!

**Comment by pokey** [comment]
Fwiw I built a [basic VSCode extension](https://github.com/pokey/git-branchless-vscode) that I've been using for a couple months now.  At some point I'll publish it, add some docs, record a video, etc, but figured I'd share it now just in case anyone wants to build off of it.  Might provide useful scaffolding for building something more full-featured

**Comment by pokey** [comment]
Here's [a video](https://youtu.be/E-8Jyok4BYE) taking the VSCode extension for a spin

**Reply by arxanas** [reply]
Thanks for sharing, very cool! I'm curious if git-branchless can do anything to aid in accessibility/screenreader workflows given that it outputs to the terminal?

**Reply by pokey** [reply]
Good question. In Cursorless, we like to put little dots over letters (called "hats") that allows us to refer to tokens by the given letter. At some point I'd love to be able to put these hats onto the commits in the smart log, so we can refer to commits that way instead of needing to select / copy them, but putting these hats over shell output is something we haven't really solved for any command yet, so I'm not sure how much git-branchless can do at this point. I'll definitely reach out when we start investing in this direction if there's something branchless can do to make things easier here

cc/ @phillco

**Reply by arxanas** [reply]
jj prints short change IDs in its graph output by default, which is really useful to address commits that you're working on.
